Documentation/technical/obsolescence.txt

```diff
@@ -0,0 +1,497 @@
+can be found and used to bring the streams back together using the robust
+merging functionality that git provides.
+
+This method also has its weaknesses. The project history ends up being far more
+complicated to read because a single logical change, meant to be merged into a
+project in a single atomic operation, comes in many commits. Even trivial
+mistakes that were easily caught in peer review, are permanently stored in the
+project's mainline history. It can be like a minefield of red herrings that can
+easily trip up `git bisect` operations and result in endless "how did this ever
+work" moments where you're going through the git log and you don't realize that
+a problem with a commit you're reading was never actually a real problem because
+it never really existed outside of development.
+
+The Rebase Method
+-----------------
+
+In the rebase method, rewriting history is the way to go. When the change itself
+needs to be modified, an author amends the existing commit, destroying the
+original one. To keep in sync with the rest of the project, an author rebases
+it. This essentially goes through the same process as merging but the result is
+a new amended commit that obsoletes, and forgets, the old one. In both cases,
+the new commit is proposed by force-pushing to a branch.
+
+This method is strong in the way it leaves the project history with fewer
+commits, each of which represents the whole of a mature change. Git bisect has
+an easier time rooting out the origin of real problems and sorting through the
+project history commit by commit is more meaningful because it represents a more
+concise and accurate picture of the project's public history.
+
+This method also has its weaknesses. Over the course of a change, the history of
+the change is destroyed. Some tools, such as Gerrit, attempt to save this
+history by attaching an immutable change id which is tracked as the change is
+amended and resubmitted. However, it is linear and limited. It breaks down when
+a change begins to evolve in multiple divergent streams of effort. For example,
+a reviewer might see an obvious error and attempt to correct it on the spot. Or,
+the reviewer might find it easier to suggest a code change explicitly rather
+than try to explain what they want. If the author is working on something else
+in the meantime, they can easily, and undetectably, clobber these suggestions in
+their next push.
+
+The common base between these streams is continually destroyed by amending
+history and it can become very difficult to make sense of it. The tool cannot
+tell where things diverged and therefore cannot tell if something is getting
+clobbered. If a larger change is broken into multiple commits, this job becomes
+even more difficult and confusing.
+
+The Obsolescence Method
+-----------------------
+
+This proposal is to keep the development history of a proposed change as a
+second orthogonal dimension of history.
```

Comment:
I should mention here that this method embraces the history rewriting mechanisms of the rebase method mentioned above. I should also mention that obsolescence supports a number of different workflows that use history rewriting.